### ISSUE TITLE
Improve multi threading in archiving to be more linear

### ISSUE THREAD

**Issue by mgazdzik** [piwik/piwik#5363]
Currently we can trigger manually many core:archive and each of them would trigger separate process for archiving idsites using common queue. However, each of those commands can spawn up to 3 (or more - if changed in file) processes computing segments. This causes that we cannot trigger as many archiving processes for idsites as we would like, because in worst case we will end up having 3x more processes computing segmented data at the same time. Therefore it would be good to have common limit saying how many process total can be spawned. Basically it should work the same for idsites as it works for segments now. That way we would be able to set top limit of process number regardless it's working on idsites of segments. Also it would allow easier managing number of processes - instead of multiple lines in crontab, only change single param to increase number of threads.

**Comment by mattab** [comment]
I guess here what we need is a simple new plugin that implements a Job queue for certain `climulti:request` calls. This is kinda similar to how [QueuedTracking](https://github.com/piwik/plugin-QueuedTracking) has a Redis queue with pending requests.

**Comment by mnapoli** [comment]
@mattab I don't understand how the work queue is going to help? I think I don't understand the issue actually, from what I could gather from the ticket the problem is because of too many threads/not enough control on those threads?

**Comment by mattab** [comment]
Right now each core:archive script will trigger 1-N new calls and there could be already some running in the background, which could overload the server.

Having a  queue (eg. FIFO) gives us ability to decide process 1 or N jobs at a time (at most) giving us a controlled environment.

**Comment by mnapoli** [comment]
Thanks I see the point of the queue now.

> each core:archive script will trigger 1-N new calls and there could be already some running in the background

Is it because core:archive doesn't wait for the requests to finish processing? Or is it because we can run multiple core:archive in parallel (but in that case this is a problem we create ourselves)? Or is there another reason why "there could be already some running in the background" (assuming archiving in the browser is disabled)?

**Comment by mattab** [comment]
there are several reasons this could happen, for exahmple if users add 100 sites overnight, or many segments (see also https://github.com/piwik/piwik/issues/7483 )... 

it is by design that user can trigger several archiving scripts, eg. to make better use of multiple CPUs on the server it could be useful. When a script is already running (for example it could run for days), if we block other core:archive calls, then data for 'yesterday', 'today', etc. may be missing. This would not be good solution

**Comment by mattab** [comment]
Not needed for now
